Source/Core/InputCommon/ControllerEmu/StickGate.cpp

```diff
@@ -46,34 +91,145 @@ ControlState SquareStickGate::GetRadiusAtAngle(double ang) const
+  const u32 sample1_index = u32(sample_pos) % m_calibration.size();
+  const u32 sample2_index = (sample1_index + 1) % m_calibration.size();
+  const double sample1_ang = sample1_index * MathUtil::TAU / m_calibration.size();
+  const double sample2_ang = sample2_index * MathUtil::TAU / m_calibration.size();
+
+  return GetRayLineIntersection(
+      GetPointFromAngleAndLength(ang, 1.0),
+      GetPointFromAngleAndLength(sample1_ang, m_calibration[sample1_index]),
+      GetPointFromAngleAndLength(sample2_ang, m_calibration[sample2_index]));
+}
+
+ControlState ReshapableInput::GetDefaultInputRadiusAtAngle(double ang) const
+{
+  // This will normally be the same as the gate radius.
+  // Unless a sub-class is doing weird things with the gate radius (e.g. Tilt)
+  return GetGateRadiusAtAngle(ang);
 }
 
-void ReshapableInput::AddReshapingSettings(ControlState default_radius, ControlState default_shape,
-                                           int max_deadzone)
+void ReshapableInput::SetCalibrationToDefault()
 {
-  // Allow radius greater than 1.0 for definitions of rounded squares
-  // This is ideal for Xbox controllers (and probably others)
-  numeric_settings.emplace_back(
-      std::make_unique<NumericSetting>(_trans("Input Radius"), default_radius, 0, 140));
-  numeric_settings.emplace_back(
-      std::make_unique<NumericSetting>(_trans("Input Shape"), default_shape, 0, 50));
-  numeric_settings.emplace_back(std::make_unique<NumericSetting>(_trans("Dead Zone"), 0, 0, 50));
+  m_calibration.clear();
+}
+
+void ReshapableInput::SetCalibrationToZero()
+{
+  m_calibration.assign(CALIBRATION_SAMPLE_COUNT, 0.0);
+}
+
+void ReshapableInput::SetCalibrationFromGate(const StickGate& gate)
+{
+  m_calibration.resize(gate.GetIdealCalibrationSampleCount().value_or(CALIBRATION_SAMPLE_COUNT));
+
+  u32 i = 0;
+  for (auto& val : m_calibration)
+    val = gate.GetRadiusAtAngle(MathUtil::TAU * i++ / m_calibration.size());
+}
+
+void ReshapableInput::UpdateCalibration(ControlState x, ControlState y)
+{
+  const auto ang_scale = MathUtil::TAU / m_calibration.size();
+
+  const u32 calibration_index =
+      std::lround((std::atan2(y, x) + MathUtil::TAU) / ang_scale) % m_calibration.size();
+  const double calibration_ang = calibration_index * ang_scale;
+  auto& calibration_sample = m_calibration[calibration_index];
+
+  // Update closest sample from provided x,y.
+  calibration_sample = std::max(calibration_sample, Common::DVec2(x, y).Length());
+
+  // Here we update all other samples in our calibration vector to maintain
+  // a convex polygon containing our new calibration point.
+  // This is required to properly fill in angles that cannot be gotten.
+  // (e.g. Keyboard input only has 8 possible angles)
+
+  // Note: Loop assumes an even sample count, which should not be a problem.
+  for (auto sample_offset = u32(m_calibration.size() / 2 - 1); sample_offset > 1; --sample_offset)
+  {
+    const auto update_at_offet = [&](u32 offset1, u32 offset2) {
```

Comment:
I _think_ you meant to spell out `offset` here instead of `offet`

Reply:
Yes. Thanks.